src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -4355,47 +4364,54 @@ private static TypeWithState GetNullCoalescingResultType(TypeWithState rightResu
+                    Debug.Assert(node is BoundUnconvertedConditionalOperator);
+                    Debug.Assert(_returnTypesOpt is not null);
+                    resultState = default;
                 }
-
-                if (alternative != originalAlternative)
+                else
                 {
-                    TrackAnalyzedNullabilityThroughConversionGroup(resultTypeWithState, (BoundConversion)originalAlternative, alternative);
+                    var resultTypeWithAnnotations = TypeWithAnnotations.Create(resultType);
+
+                    TypeWithState convertedConsequenceResult = ConvertConditionalOperandOrSwitchExpressionArmResult(
+                        originalConsequence,
+                        consequence,
+                        consequenceConversion,
+                        resultTypeWithAnnotations,
+                        consequenceRValue,
+                        consequenceState,
+                        consequenceEndReachable);
+
+                    TypeWithState convertedAlternativeResult = ConvertConditionalOperandOrSwitchExpressionArmResult(
+                        originalAlternative,
+                        alternative,
+                        alternativeConversion,
+                        resultTypeWithAnnotations,
+                        alternativeRValue,
+                        alternativeState,
+                        alternativeEndReachable);
+
+                    resultState = convertedConsequenceResult.State.Join(convertedAlternativeResult.State);
                 }
             }
             else
             {
-                var resultTypeWithAnnotations = TypeWithAnnotations.Create(resultType);
-
-                TypeWithState convertedConsequenceResult = convertResult(
-                    originalConsequence,
-                    consequence,
-                    consequenceConversion,
-                    resultTypeWithAnnotations,
-                    consequenceRValue,
-                    consequenceState,
-                    consequenceEndReachable);
-
-                TypeWithState convertedAlternativeResult = convertResult(
-                    originalAlternative,
-                    alternative,
-                    alternativeConversion,
-                    resultTypeWithAnnotations,
-                    alternativeRValue,
-                    alternativeState,
-                    alternativeEndReachable);
+                resultState = consequenceRValue.State.Join(alternativeRValue.State);
 
-                resultState = convertedConsequenceResult.State.Join(convertedAlternativeResult.State);
+                _conditionalInfoForConversionOpt ??= PooledDictionary<BoundExpression, ImmutableArray<(LocalState, TypeWithState, bool)>>.GetInstance();
+                _conditionalInfoForConversionOpt.Add(node, ImmutableArray.Create(
+                    (consequenceState, consequenceRValue, consequenceEndReachable),
```

Comment:
>consequenceState [](start = 21, length = 16)

I think `consequenceState` is the state *after* visiting the consequence when we reach here. But I expected that we'd store the state *before* visiting the consequence, based on the comment on `_conditionalInfoForConversionOpt`.

> This is then used by VisitConversion to properly set the state *before* each branch when visiting a conversion applied to such a construct.

```
// line 4334 above
                (consequence, consequenceConversion, consequenceRValue) = visitConditionalOperand(consequenceState, originalConsequence);
                Unsplit();
                consequenceState = this.State;
```

Reply:
It depends. If the nested operator was a lambda, it's unchanged from before the visit operand, because the lambda was skipped. If it isn't a lambda, then it's the state after visiting, which is still correct because that's the state that should be set when we call `VisitConversion` for real on the operand.

---
In reply to: [595627389](https://github.com/dotnet/roslyn/pull/51791#discussion_r595627389) [](ancestors = 595627389)